internal/uvm/vsmb.go

```diff
@@ -32,31 +34,54 @@ func (uvm *UtilityVM) AddVSMB(ctx context.Context, hostPath string, guestRequest
 
 	uvm.m.Lock()
 	defer uvm.m.Unlock()
-	share, err := uvm.findVSMBShare(ctx, hostPath)
+
+	st, err := os.Stat(hostPath)
+	if err != nil {
+		return err
+	}
+	var file string
+	m := uvm.vsmbDirShares
+	if !st.IsDir() {
+		m = uvm.vsmbFileShares
+		file = hostPath
+		hostPath = filepath.Clean(filepath.Dir(hostPath))
+		options.RestrictFileAccess = true
+		options.SingleFileMapping = true
+	}
+	var requestType = requesttype.Update
+	share, err := uvm.findVSMBShare(ctx, m, hostPath)
 	if err == ErrNotAttached {
+		requestType = requesttype.Add
 		uvm.vsmbCounter++
 		shareName := "s" + strconv.FormatUint(uvm.vsmbCounter, 16)
 
+		share = &vsmbShare{
+			name:         shareName,
+			guestRequest: guestRequest,
+		}
+	}
+	allowedFiles := append(share.allowedFiles, file)
+
+	if requestType == requesttype.Add || len(allowedFiles) > 0 {
```

Comment:
I'm struggling to understand this logic. Will this not always be 1 and tied to requestttype.Update?

Reply:
I added some comments (and a thorough commit message), can you see if they answer your question? I want to make sure it's clear. :)